### ISSUE TITLE
ocsp update over cli crashing haproxy

### ISSUE THREAD

**Issue by pa77777** [haproxy/haproxy#2442]
### Detailed Description of the Problem

We are loading certificates from directory, so we are updating ocsp over cli, we have lot of certificates (about 500), sending one ocsp update after another in bash script with zero size of .ocsp files
<code>
		echo -e "set ssl cert ${CERT}.ocsp <<\n$(cat ${CERT}.ocsp|awk '!/^$/')\n"|socat stdio unix-connect:${SOCKET} \
                && echo -e "commit ssl cert ${CERT}"|socat stdio unix-connect:${SOCKET} \
		&& echo -e "update ssl ocsp-response ${CERT}"|socat stdio unix-connect:${SOCKET}
</code>
then we see lot of records this type in logs, without ocsp update there is no any segfault
<code>
kernel: haproxy[1914748]: segfault at 10 ip 00000000005b20e8 sp 00007fdbb91f2690 error 4 in haproxy[400000+60a000]
</code>
I improved this situation with sleep 0.x seconds between each update but still it crashes

### Expected Behavior

not crashing–;) and give ocsp-update to bind line for ssl from directory

### Steps to Reproduce the Behavior

repeat this command on lot of certs, give it few hours and it crashes
<code>
		echo -e "set ssl cert ${CERT}.ocsp <<\n$(cat ${CERT}.ocsp|awk '!/^$/')\n"|socat stdio unix-connect:${SOCKET} \
                && echo -e "commit ssl cert ${CERT}"|socat stdio unix-connect:${SOCKET} \
		&& echo -e "update ssl ocsp-response ${CERT}"|socat stdio unix-connect:${SOCKET}
</code>

### Do you have any idea what may have caused this?

cli or ocsp update ? more updates, more often crashes 

### What is your configuration?

```haproxy
bind ipv4@"$IPV4_ADDRESS":443,ipv6@"$IPV6_ADDRESS":443 strict-sni ssl crt /etc/nethostssl alpn h2,http/1.1
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.8.5-aaba8d0 2023/12/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.5.html
Running on: Linux 4.18.0-372.26.1.el8_6.x86_64 #1 SMP Tue Sep 13 18:09:48 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN -SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 1.1.1w+quic  11 Sep 2023
Running on OpenSSL version : OpenSSL 1.1.1w+quic  11 Sep 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.5.0 20210514 (Red Hat 8.5.0-20)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
1. kernel: haproxy[1914748]: segfault at 10 ip 00000000005b20e8 sp 00007fdbb91f2690 error 4 in haproxy[400000+60a000]
2. Thread 4 (Thread 0x7fdba51ff700 (LWP 1914749)):
#0  0x00007fdbbb55ea07 in epoll_wait () from /lib64/libc.so.6
No symbol table info available.
#1  0x000000000047310f in _do_poll (p=<optimized out>, exp=<optimized out>, wake=<optimized out>) at src/ev_epoll.c:232
        timeout = 68
        status = <optimized out>
        fd = <optimized out>
        count = <optimized out>
        updt_idx = <optimized out>
        wait_time = 68
        old_fd = <optimized out>
#2  0x00000000006071f8 in run_poll_loop () at src/haproxy.c:3044
        next = <optimized out>
        wake = <optimized out>
        __func__ = <optimized out>
#3  0x00000000006078a7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3168
        ptaf = <optimized out>
        ptif = 0xc52750 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0,
              __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 13, __wseq32 = {__low = 13, __high = 0}}, {__g1_start = 7, __g1_start32 = {__low = 7, __high = 0}}, __g_refs = {0,
              0}, __g_size = {0, 0}, __g1_orig_size = 12, __wrefs = 0, __g_signals = {0, 0}},
--Type <RET> for more, q to quit, c to continue without paging--c
          __size = "\r\000\000\000\000\000\000\000\a", '\000' <repeats 23 times>, "\f", '\000' <repeats 14 times>, __align = 13}
#4  0x00007fdbbb7fc1cf in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#5  0x00007fdbbb468dd3 in clone () from /lib64/libc.so.6
No symbol table info available.

Thread 3 (Thread 0x7fdbbd6a5bc0 (LWP 1914747)):
#0  0x00007fdbbb55ea07 in epoll_wait () from /lib64/libc.so.6
No symbol table info available.
#1  0x000000000047310f in _do_poll (p=<optimized out>, exp=<optimized out>, wake=<optimized out>) at src/ev_epoll.c:232
        timeout = 26
        status = <optimized out>
        fd = <optimized out>
        count = <optimized out>
        updt_idx = <optimized out>
        wait_time = 26
        old_fd = <optimized out>
#2  0x00000000006071f8 in run_poll_loop () at src/haproxy.c:3044
        next = <optimized out>
        wake = <optimized out>
        __func__ = <optimized out>
#3  0x00000000006078a7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3168
        ptaf = <optimized out>
        ptif = 0xc52750 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 13, __wseq32 = {__low = 13, __high = 0}}, {__g1_start = 7, __g1_start32 = {__low = 7, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 12, __wrefs = 0, __g_signals = {0, 0}}, __size = "\r\000\000\000\000\000\000\000\a", '\000' <repeats 23 times>, "\f", '\000' <repeats 14 times>, __align = 13}
#4  0x0000000000470b02 in main (argc=<optimized out>, argv=<optimized out>) at src/haproxy.c:3858
        err = <optimized out>
        retry = <optimized out>
        limit = {rlim_cur = 524995, rlim_max = 524995}
        pidfd = <optimized out>
        intovf = <optimized out>
        msg = <optimized out>

Thread 2 (Thread 0x7fdba49fe700 (LWP 1914750)):
#0  0x00007fdbbb55ea07 in epoll_wait () from /lib64/libc.so.6
No symbol table info available.
#1  0x000000000047310f in _do_poll (p=<optimized out>, exp=<optimized out>, wake=<optimized out>) at src/ev_epoll.c:232
        timeout = 69
        status = <optimized out>
        fd = <optimized out>
        count = <optimized out>
        updt_idx = <optimized out>
        wait_time = 69
        old_fd = <optimized out>
#2  0x00000000006071f8 in run_poll_loop () at src/haproxy.c:3044
        next = <optimized out>
        wake = <optimized out>
        __func__ = <optimized out>
#3  0x00000000006078a7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3168
        ptaf = <optimized out>
        ptif = 0xc52750 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 13, __wseq32 = {__low = 13, __high = 0}}, {__g1_start = 7, __g1_start32 = {__low = 7, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 12, __wrefs = 0, __g_signals = {0, 0}}, __size = "\r\000\000\000\000\000\000\000\a", '\000' <repeats 23 times>, "\f", '\000' <repeats 14 times>, __align = 13}
#4  0x00007fdbbb7fc1cf in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#5  0x00007fdbbb468dd3 in clone () from /lib64/libc.so.6
No symbol table info available.

Thread 1 (Thread 0x7fdbb91ff700 (LWP 1914748)):
#0  sess_build_logline (sess=sess@entry=0x7fdbb841ce80, s=s@entry=0x0, dst=0x7fdbb8414500 "", maxsize=1024, list_format=<optimized out>) at src/log.c:2087
        conn = <optimized out>
        key = <optimized out>
        empty = {size = 504, area = 0x7fdbb6851840 "", data = 140581680392064, head = 140581319511543}
        addr = <optimized out>
        src = 0x0
        fe = <optimized out>
        be = <optimized out>
        txn = <optimized out>
        logs = <optimized out>
        fe_conn = <optimized out>
        be_conn = <optimized out>
        s_flags = <optimized out>
        uniq_id = <optimized out>
        chunk = {size = 140581635459936, area = 0x496841 <ssl_sock_load_ocsp_response+609> "H\201\304\030\001", data = 12719456, head = 503}
        uri = <optimized out>
        spc = <optimized out>
        qmark = <optimized out>
        end = <optimized out>
        tm = {tm_sec = -1225189264, tm_min = 32731, tm_hour = -1232354176, tm_mday = 32731, tm_mon = -1232300672, tm_year = 32731, tm_wday = -1189140435, tm_yday = 32731, tm_isdst = 58, tm_gmtoff = 47244640256, tm_zone = 0x7c00000001 <error: Cannot access memory at address 0x7c00000001>}
        t_request = <optimized out>
        hdr = <optimized out>
        last_isspace = <optimized out>
        nspaces = 0
        tmplog = 0x7fdbb8414500 ""
        ret = <optimized out>
        iret = <optimized out>
        status = 0
        tmp = 0x0
        tv = {tv_sec = 140581644343408, tv_usec = 0}
        tmp_strm_log = {logwait = 0, level = 0, accept_date = {tv_sec = 140581637662864, tv_usec = 799533}, accept_ts = 1709397883478898000, t_handshake = 0, t_idle = 0, request_ts = 0, t_queue = -1, t_connect = -1, t_data = -1, t_close = 4154504683, srv_queue_pos = 0, prx_queue_pos = 0, bytes_in = 0, bytes_out = 0}
        path = <optimized out>
        parser = {uri = {ptr = 0x7fdbb68b1480 "\201ߒ\266\333\177", len = 7361461}, state = 3069778032, format = 32731}
#1  0x00000000005b5c32 in sess_log (sess=0x7fdbb841ce80) at src/log.c:3159
        size = <optimized out>
        level = 3
        sd_size = 0
#2  0x00000000004978f6 in ssl_ocsp_send_log () at src/ssl_ocsp.c:1061
No locals.
#3  ssl_ocsp_update_responses (task=0x7fdba3218dc0, context=<optimized out>, state=<optimized out>) at src/ssl_ocsp.c:1156
        next_wakeup = 0
        eb = <optimized out>
        ocsp = 0x7fdbb68b1480
        hc = 0x7fdba3a8e100
        req_url = 0x0
        req_body = 0x0
        certid = 0x0
        ctx = 0xc5b980 <ssl_ocsp_task_ctx>
        __lk_r = <optimized out>
        __set_r = <optimized out>
        __msk_r = <optimized out>
        ret = <optimized out>
        __lk_r = <optimized out>
        __set_r = <optimized out>
        __msk_r = <optimized out>
        ret = <optimized out>
#4  0x0000000000639fda in run_tasks_from_lists (budgets=budgets@entry=0x7fdbb91f29a0) at src/task.c:634
        process = <optimized out>
        tl_queues = <optimized out>
        t = 0x7fdba3218dc0
        budget_mask = <optimized out>
        profile_entry = 0x0
        done = 1
        queue = <optimized out>
        state = <optimized out>
        ctx = <optimized out>
        __func__ = <optimized out>
#5  0x000000000063a9d6 in process_runnable_tasks () at src/task.c:876
        tt = 0xe71580 <ha_thread_ctx+512>
        lrq = <optimized out>
        grq = <optimized out>
        t = <optimized out>
        max = {0, 135, 0, 0}
        max_total = <optimized out>
        tmp_list = <optimized out>
        queue = <optimized out>
        max_processed = <optimized out>
        lpicked = <optimized out>
        gpicked = <optimized out>
        heavy_queued = 1
        budget = <optimized out>
        __lk_r = <optimized out>
        __set_r = <optimized out>
        __msk_r = <optimized out>
        ret = <optimized out>
#6  0x0000000000607237 in run_poll_loop () at src/haproxy.c:2969
        next = <optimized out>
        wake = <optimized out>
        __func__ = <optimized out>
#7  0x00000000006078a7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3168
        ptaf = <optimized out>
        ptif = 0xc52750 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 13, __wseq32 = {__low = 13, __high = 0}}, {__g1_start = 7, __g1_start32 = {__low = 7, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 12, __wrefs = 0, __g_signals = {0, 0}}, __size = "\r\000\000\000\000\000\000\000\a", '\000' <repeats 23 times>, "\f", '\000' <repeats 14 times>, __align = 13}
#8  0x00007fdbbb7fc1cf in start_thread () from /lib64/libpthread.so.0
No symbol table info available.
#9  0x00007fdbbb468dd3 in clone () from /lib64/libc.so.6
No symbol table info available.
```

**Comment by wtarreau** [comment]
Thank you. Can you please issue "where" in gdb so that we know precisely where it crashed ? It mentions line 2087 in log.c but that's a variable declaration so I suspect it's in fact the next line with switch(tmp->type). But if so, it would mean that a list entry found in the list is wrong or possibly corrupted, which is quite confusing. Maybe please also issue "p *tmp" in gdb, in case it can print something about this strange entry.

**Comment by wtarreau** [comment]
Oh I'm seeing tmp=0 in your back trace. That's definitely not good. I don't see how it can happen but suspect the list itself is null. I'm investigating.

**Comment by wlallemand** [comment]
I think that's the problem we already discovered with @rlebreton a while ago, there are a few problems when combining ocsp-update and the CLI currently.

**Comment by wtarreau** [comment]
Apparently it died when exploiting the log-format-sd, is this defined in your config ?

**Comment by wtarreau** [comment]
I'd be interested in showing the proxy's contents, e.g. "p *sess->fe". I'm seeing how this could cause this if sess->fe->logformat_sd is NULL, except that my understanding of the code is that this proxy is the httpclient's and the httpclient uses the standard alloc_new_proxy() call which performs a LIST_INIT() on that field. So either I'm missing something, or it got overwritten at some point, I don't know.

**Comment by wtarreau** [comment]
So I verified in gdb and the proxy doesn't have a NULL logformat_sd list. Hence it must have been destroyed later (maybe a free before detaching an element for example).

**Comment by wlallemand** [comment]
I pushed @rlebreton fixes related to 'set ssl cert'  + 'ocsp-update', it's possible that this problem is related.

**Comment by pa77777** [comment]
after last patch 2.8.6 I see some fixes on ocsp in changlogs
```
    - BUG/MINOR: ssl: Duplicate ocsp update mode when dup'ing ckch
    - MINOR: ssl: Use OCSP_CERTID instead of ckch_store in ckch_store_build_certid
    - BUG/MEDIUM: ocsp: Separate refcount per instance and per store
    - BUG/MINOR: ssl: Reenable ocsp auto-update after an "add ssl crt-list"
    - REGTESTS: ssl: Add OCSP related tests
    - BUG/MEDIUM: ssl: Fix crash when calling "update ssl ocsp-response" when an update is ongoing
```
but when I send `update ssl ocsp-response` haproxy crashes totaly
```
haproxy[2501588]: FATAL: bug condition "ocsp->refcount_instance > 0" matched at src/ssl_ocsp.c:397
haproxy[2501588]:   call trace(15):
haproxy[2501588]:   |       0x4953b6 [0f 0b 0f 1f 84 00 00 00]: main+0x26376
haproxy[2501588]:   |       0x48468b [48 8b 7b 38 e8 cc 80 2e]: ssl_sock_free_cert_key_and_chain_contents+0x16b/0x1c7
haproxy[2501588]:   |       0x485143 [48 8b 7d 00 e8 c4 80 fe]: ckch_store_free+0x53/0x71
haproxy[2501588]:   |       0x488309 [5b 48 8d 75 28 bf 90 cd]: ckch_store_replace+0x119/0x12f
haproxy[2501588]:   |       0x48838f [48 c7 03 00 00 00 00 48]: main+0x1934f
haproxy[2501588]:   |       0x5db763 [85 c0 74 22 49 8b 45 40]: main+0x16c723
haproxy[2501588]:   |       0x68868c [44 8b 05 35 9f 5c 00 45]: task_run_applet+0x13c/0xdd3
haproxy[2501588]:   |       0x63b07a [44 0f b6 44 24 18 48 89]: run_tasks_from_lists+0x38a/0x9fa
haproxy[2501588]:   |       0x63ba76 [29 44 24 1c 8b 4c 24 1c]: process_runnable_tasks+0x386/0x6f7
haproxy[2501588]:   |       0x608237 [83 3d 22 9c 8b 00 01 0f]: run_poll_loop+0x127/0x5a8
haproxy[2501588]:   |       0x6088a7 [48 8b 1d 92 ae 64 00 48]: main+0x199867
haproxy[2501588]:   | 0x7f1968c401ca [64 48 89 04 25 30 06 00]: libpthread:+0x81ca
haproxy[2501588]:   | 0x7f19688ace73 [48 89 c7 b8 3c 00 00 00]: libc:clone+0x43/0x5e
haproxy[2501577]: [NOTICE]   (2501577) : haproxy version is 2.8.6-f6bd011
haproxy[2501577]: [NOTICE]   (2501577) : path to executable is /usr/local/haproxy/sbin/haproxy
haproxy[2501577]: [ALERT]    (2501577) : Current worker (2501588) exited with code 132 (Illegal instruction)
haproxy[2501577]: [ALERT]    (2501577) : exit-on-failure: killing every processes with SIGTERM
haproxy[2501577]: [WARNING]  (2501577) : All workers exited. Exiting... (132)
```

**Comment by wtarreau** [comment]
Many thanks for this. As I mentioned in the 3.0-dev4 announce, we had a somewhat similar report but were unable to reproduce it. Probably that your backtrace will be helpful!

**Comment by delagen** [comment]
I have crash on start with ocsp error when binary was compiled with libressl. Recompiled with openssl, and error was gone

**Comment by wtarreau** [comment]
OK that's useful to know as well, thanks!

**Comment by pa77777** [comment]
```
       Message: Process 1867936 (haproxy) of user 992 dumped core.

                Stack trace of thread 1867937:
                #0  0x000000000049504e ssl_sock_free_ocsp (haproxy + 0x9504e)
                #1  0x00000000004841a8 ssl_sock_free_cert_key_and_chain_contents (haproxy + 0x841a8)
                #2  0x0000000000484c23 ckch_store_free (haproxy + 0x84c23)
                #3  0x0000000000487e90 ckch_store_replace (haproxy + 0x87e90)
                #4  0x0000000000487f2f cli_io_handler_commit_cert (haproxy + 0x87f2f)
                #5  0x00000000005d14d7 cli_io_handler (haproxy + 0x1d14d7)
                #6  0x0000000000677df7 task_run_applet (haproxy + 0x277df7)
                #7  0x000000000062e5d5 run_tasks_from_lists (haproxy + 0x22e5d5)
                #8  0x000000000062ef09 process_runnable_tasks (haproxy + 0x22ef09)
                #9  0x00000000005fc757 run_poll_loop (haproxy + 0x1fc757)
                #10 0x00000000005fcdc7 run_thread_poll_loop (haproxy + 0x1fcdc7)
                #11 0x00007f1475a9f802 start_thread (libc.so.6 + 0x9f802)
                #12 0x00007f1475a3f450 __clone3 (libc.so.6 + 0x3f450)

                Stack trace of thread 1867936:
                #0  0x00007f1475b4e84e epoll_wait (libc.so.6 + 0x14e84e)
                #1  0x0000000000473261 _do_poll (haproxy + 0x73261)
                #2  0x00000000005fc718 run_poll_loop (haproxy + 0x1fc718)
                #3  0x00000000005fcdc7 run_thread_poll_loop (haproxy + 0x1fcdc7)
                #4  0x00000000004716df main (haproxy + 0x716df)
                #5  0x00007f1475a3feb0 __libc_start_call_main (libc.so.6 + 0x3feb0)
                #6  0x00007f1475a3ff60 __libc_start_main@@GLIBC_2.34 (libc.so.6 + 0x3ff60)
                #7  0x0000000000471f15 _start (haproxy + 0x71f15)
                ELF object binary architecture: AMD x86-64
                

(gdb) t a a bt full

Thread 2 (Thread 0x7f1475cc1a40 (LWP 1867936)):
#0  0x00007f1475b4e84e in epoll_wait () from /lib64/libc.so.6
No symbol table info available.
#1  0x0000000000473261 in _do_poll (p=<optimized out>, exp=<optimized out>, wake=<optimized out>) at src/ev_epoll.c:232
        timeout = 132
        status = <optimized out>
        fd = <optimized out>
        count = <optimized out>
        updt_idx = <optimized out>
        wait_time = 132
        old_fd = <optimized out>
#2  0x00000000005fc718 in run_poll_loop () at src/haproxy.c:3045
        next = <optimized out>
        wake = <optimized out>
        __func__ = {<optimized out> <repeats 14 times>}
#3  0x00000000005fcdc7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3169
        ptaf = <optimized out>
        ptif = 0xa42750 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
--Type <RET> for more, q to quit, c to continue without paging--c
        init_cond = {__data = {{__wseq = 3, __wseq32 = {__low = 3, __high = 0}}, {__g1_start = 1, __g1_start32 = {__low = 1, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 4, __wrefs = 0, __g_signals = {0, 0}}, __size = "\003\000\000\000\000\000\000\000\001", '\000' <repeats 23 times>, "\004", '\000' <repeats 14 times>, __align = 3}
#4  0x00000000004716df in main (argc=<optimized out>, argv=<optimized out>) at src/haproxy.c:3859
        err = <optimized out>
        retry = <optimized out>
        limit = {rlim_cur = 65535, rlim_max = 65535}
        pidfd = <optimized out>
        intovf = <optimized out>
        msg = <optimized out>

Thread 1 (Thread 0x7f14751ff640 (LWP 1867937)):
#0  ssl_sock_free_ocsp (ocsp=0x7f147424ad00) at src/ssl_ocsp.c:397
No locals.
#1  0x000000000049686a in ssl_sock_free_ocsp (ocsp=<optimized out>) at src/ssl_ocsp.c:391
        __lk_r = <optimized out>
        __set_r = <optimized out>
        __msk_r = <optimized out>
        ret = <optimized out>
        __x = <optimized out>
        __x = <optimized out>
#2  0x00000000004841a8 in ssl_sock_free_cert_key_and_chain_contents (data=0x7f147562de30) at src/ssl_ckch.c:738
        ocsp = <optimized out>
        certid = "0K0\t\006\005+\016\003\002\032\005\000\004\024H\332ɠ\373+\323-O\360\336h\322\365g\267\065\371\263\304\004\024\024.\263\027\267XVˮP\t@\346\037\257\235\213\024\302\306\002\022\004w\030\315\346\035\356Y|\006\214\025\367\355\260\210\345\n", '\000' <repeats 50 times>
        certid_length = 77
#3  0x0000000000484c23 in ckch_store_free (store=0x7f1475645730) at src/ssl_ckch.c:896
        inst = <optimized out>
        inst_s = 0x7f14756456f0
#4  0x0000000000487e90 in ckch_store_replace (old_ckchs=0x7f1475645730, new_ckchs=new_ckchs@entry=0x7f147439a610) at src/ssl_ckch.c:2124
        entry = <optimized out>
        ckchi = <optimized out>
        ckchis = <optimized out>
#5  0x0000000000487f2f in cli_io_handler_commit_cert (appctx=0x7f1474401280) at src/ssl_ckch.c:2211
        ctx = 0x7f1474401330
        sc = <optimized out>
        y = 8
        old_ckchs = <optimized out>
        new_ckchs = <optimized out>
        ckchi = <optimized out>
#6  0x00000000005d14d7 in cli_io_handler (appctx=0x7f1474401280) at src/cli.c:1115
        ctx = <optimized out>
        msg = <optimized out>
        sev = <optimized out>
        sc = 0x7f1474496060
        req = <optimized out>
        res = 0x7f147449d070
        bind_conf = 0x7f14741f6380
        reql = <optimized out>
        len = <optimized out>
        __func__ = {<optimized out> <repeats 15 times>}
#7  0x0000000000677df7 in task_run_applet (t=0x7f1474419320, context=<optimized out>, state=<optimized out>) at src/applet.c:454
        app = <optimized out>
        sc = <optimized out>
        sco = 0x7f1474496000
        rate = <optimized out>
        count = 46
        did_send = 0
        __FUNCTION__ = "task_run_applet"
#8  0x000000000062e5d5 in run_tasks_from_lists (budgets=budgets@entry=0x7f14751f2a80) at src/task.c:634
        process = <optimized out>
        tl_queues = <optimized out>
        t = 0x7f1474419320
        budget_mask = <optimized out>
        profile_entry = 0x0
        done = <optimized out>
        queue = <optimized out>
        state = <optimized out>
        ctx = <optimized out>
        __func__ = {<optimized out> <repeats 21 times>}
#9  0x000000000062ef09 in process_runnable_tasks () at src/task.c:876
        tt = 0xc87000 <ha_thread_ctx+512>
        lrq = <optimized out>
        grq = <optimized out>
        t = <optimized out>
        max = {0, 184, 0, 0}
        max_total = <optimized out>
        tmp_list = <optimized out>
        queue = <optimized out>
        max_processed = <optimized out>
        lpicked = <optimized out>
        gpicked = <optimized out>
        heavy_queued = 1
        budget = <optimized out>
        not_done_yet = <optimized out>
#10 0x00000000005fc757 in run_poll_loop () at src/haproxy.c:2970
        next = <optimized out>
        wake = <optimized out>
        __func__ = {<optimized out> <repeats 14 times>}
#11 0x00000000005fcdc7 in run_thread_poll_loop (data=<optimized out>) at src/haproxy.c:3169
        ptaf = <optimized out>
        ptif = 0xa42750 <per_thread_init_list>
        ptdf = <optimized out>
        ptff = <optimized out>
        init_left = 0
        init_mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}
        init_cond = {__data = {{__wseq = 3, __wseq32 = {__low = 3, __high = 0}}, {__g1_start = 1, __g1_start32 = {__low = 1, __high = 0}}, __g_refs = {0, 0}, __g_size = {0, 0}, __g1_orig_size = 4, __wrefs = 0, __g_signals = {0, 0}}, __size = "\003\000\000\000\000\000\000\000\001", '\000' <repeats 23 times>, "\004", '\000' <repeats 14 times>, __align = 3}
#12 0x00007f1475a9f802 in start_thread () from /lib64/libc.so.6
No symbol table info available.
#13 0x00007f1475a3f450 in clone3 () from /lib64/libc.so.6
No symbol table info available.
```

**Comment by wtarreau** [comment]
Please note that 2.9.6 and 2.8.7 were both emitted a few hours ago to address the regression above. For now they just revert the fix as it used to affect more users than the initial issue (basically all those touching OCSP from the CLI instead of just those with the ocsp-update mechanism).

**Comment by rlebreton** [comment]
I managed to reproduce (by happy accident) the first crash you had in sess_build_logline. It's still a bit puzzling, I don't have a fix yet but at least I have a reproducer. For info, I just call the "update ssl ocsp-response" constantly and it ends up crashing.

**Comment by capflam** [comment]
@rlebreton, what is the status of this issue ? Is it fixed ?

**Comment by wlallemand** [comment]
It's supposed to be fixed with the mentioned commit indeed.

**Comment by pa77777** [comment]
yes, I think it's fixed at 2.8.9 version

**Comment by capflam** [comment]
Ok, thanks. I'm closing !
